### PR TITLE
[Feature] Delete nodes/edges.

### PR DESCRIPTION
**Pull request by yzh119**
## Description
This is an in-efficient implementation of removing nodes/edges in `DGLGraph`. As in issue #467 .

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

## Changes
In this PR, I use node/edge subgraph to implement nodes/edges removal.
However, the behavior of edge subgraph is different from edge removal. Because in edge subgraph, nodes without in/out edges would be removed, but it's node the expected behavior of edge removal.

One solution is to add an argument `preserve_nodes` in `EdgeSubgraph` function to indicate whether to preserve nodes in the graph.
Do you have any suggestions?

### CODE REVIEW

File: python/dgl/graph.py [hunk 1]
```diff
@@ -1101,6 +1101,64 @@ def add_edges(self, u, v, data=None):
...
+            self._node_frame = FrameRef(Frame(self._node_frame[sgi.induced_nodes]))
+        else:
+            self._node_frame = FrameRef(self._node_frame, sgi.induced_nodes)
+
+        if isinstance(self._edge_frame, FrameRef):
+            self._edge_frame = FrameRef(Frame(self._edge_frame[sgi.induced_edges]))
+        else:
+            self._edge_frame = FrameRef(self._edge_frame, sgi.induced_edges)
+
+        # FIXME(zihao): the following impl is not efficient, but directly reusing the handle of subgraph index would
```
Comment (yzh119) [inline python/dgl/graph.py]: @jermainewang and @zheng-da , could you please have a look at this? After discussion with @BarclayII we think the problem resides in "GraphInterface.Reset".

Comment (yzh119) [inline python/dgl/graph.py]: @jermainewang and @zheng-da could you please have a look at this, after discussion with @BarclayII we think the problem resides in "GraphInterface.Reset".

File: src/graph/graph_apis.cc [hunk 2]
```diff
@@ -452,7 +452,8 @@ DGL_REGISTER_GLOBAL("graph_index._CAPI_DGLGraphEdgeSubgraph")
     GraphHandle ghandle = args[0];
     const GraphInterface *gptr = static_cast<GraphInterface*>(ghandle);
     const IdArray eids = args[1];
-    *rv = ConvertSubgraphToPackedFunc(gptr->EdgeSubgraph(eids));
+    bool preserve_nodes = args.size() > 2 ? args[2] : false;
```
Comment (zheng-da) [inline src/graph/graph_apis.cc]: the number of arguments can change? from python, you always pass 3 arguments.

File: src/graph/immutable_graph.cc [hunk 3]
```diff
@@ -640,7 +641,8 @@ Subgraph ImmutableGraph::VertexSubgraph(IdArray vids) const {
                   sg.induced_vertices, sg.induced_edges};
 }
 
-Subgraph ImmutableGraph::EdgeSubgraph(IdArray eids) const {
+Subgraph ImmutableGraph::EdgeSubgraph(IdArray eids, bool preserve_nodes) const {
+  // TODO(zihao): to handle the case of preserve_nodes=true.
```
Comment (zheng-da) [inline src/graph/immutable_graph.cc]: maybe you can add an assertion here and above.

File: tests/compute/test_removal.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    g.del_edges(range(10, 20))
+    assert g.number_of_nodes() == 10
+    assert g.number_of_edges() == 48
+
+if __name__ == '__main__':
+    test_node_removal()
+    test_edge_removal()
+    test_multigraph_node_removal()
+    test_multigraph_edge_removal()
+    test_node_and_edge_removal()
```
Comment (zheng-da) [inline tests/compute/test_removal.py]: please check the data in the node frame and edge frame as well.

File: python/dgl/graph.py (hunk 1 above)
Comment (jermainewang) [reply python/dgl/graph.py]: You cannot let two GraphIndex object share the same internal handle. You will get double free problem. The reason is that there is no refcount for the handle (which is essentially a pointer to a C object). We rely on python's native object refcount to delete the handle properly. Checkout L37-40 in `graph_index.py`. The handle is deleted in the `__del__` function, which will be called **only** when the GraphIndex object is destroyed. Python makes sure that if a GraphIndex object is shallow-copied to another variable, only one `__del__` function call will be made.

If you create a new GraphIndex and reuse the internal handle of an existing GraphIndex object, two `__del__` function will delete the same handle.

Comment (jermainewang) [reply python/dgl/graph.py]: My solution:
* Change `graph_index.SubgraphIndex` to **not** inherit `GraphIndex`. Actually, our C++ data structure is like this. The subgraph index only contains the graph structure, the parent graph, induced nodes and edges. As a result, the object returned by `node_subgraph` is not necessarily readonly graph or not.
* In `del_nodes/edges` functions, directly set the returned graph structure as the `_graph` field in the `DGLGraph` object.

File: python/dgl/graph.py [hunk 5]
```diff
@@ -1101,6 +1101,57 @@ def add_edges(self, u, v, data=None):
             self._msg_index = self._msg_index.append_zeros(num)
         self._msg_frame.add_rows(num)
 
+    def del_nodes(self, vids):
```
Comment (jermainewang) [inline python/dgl/graph.py]: Use `remove_nodes` and `remove_edges` to be consistent with networkx.

File: python/dgl/graph.py [hunk 6]
```diff
@@ -1101,6 +1101,57 @@ def add_edges(self, u, v, data=None):
...
+            self._node_frame = FrameRef(self._node_frame, sgi.induced_nodes)
+
+        if isinstance(self._edge_frame, FrameRef):
+            self._edge_frame = FrameRef(Frame(self._edge_frame[sgi.induced_edges]))
+        else:
+            self._edge_frame = FrameRef(self._edge_frame, sgi.induced_edges)
+
+        self._graph = sgi.graph
+
+    def del_edges(self, eids):
```
Comment (jermainewang) [inline python/dgl/graph.py]: same

File: python/dgl/graph_index.py [hunk 7]
```diff
@@ -536,23 +537,26 @@ def node_subgraphs(self, vs_arr):
...
-    def edge_subgraph(self, e):
+    def edge_subgraph(self, e, preserve_nodes=False):
         """Return the induced edge subgraph.
 
         Parameters
         ----------
         e : utils.Index
             The edges.
+        preserve_nodes: bool
+            Indicates whether to preserve all nodes or not. (when set to false, all nodes without edge connected to it would be removed.)
```
Comment (jermainewang) [inline python/dgl/graph_index.py]: If true, keep the nodes which have no edge connected in the subgraph.
It is equivalent to a sparser graph that contains only the given edges.

File: python/dgl/subgraph.py [hunk 8]
```diff
@@ -32,35 +32,31 @@ class DGLSubGraph(DGLGraph):
...
+        Internal subgraph data structure.
     shared : bool, optional
         Whether the subgraph shares node/edge features with the parent graph.
     """
-    def __init__(self, parent, parent_nid, parent_eid, graph_idx, shared=False):
-        super(DGLSubGraph, self).__init__(graph_data=graph_idx,
-                                          readonly=graph_idx.is_readonly())
+    def __init__(self, parent, sgi, shared=False):
+        super(DGLSubGraph, self).__init__(graph_data=sgi.graph,
+                                          readonly=sgi.graph.is_readonly())
```
Comment (jermainewang) [inline python/dgl/subgraph.py]: `readonly=True`

File: src/graph/graph.cc [hunk 9]
```diff
@@ -467,37 +467,56 @@ Subgraph Graph::VertexSubgraph(IdArray vids) const {
...
+        {static_cast<int64_t>(nodes.size())}, eids->dtype, eids->ctx);
+    std::copy(nodes.begin(), nodes.end(), static_cast<int64_t*>(rst.induced_vertices->data));
+  } else {
+    rst.graph = std::make_shared<Graph>(IsMultigraph());
+    rst.induced_edges = eids;
+    rst.graph->AddVertices(NumVertices());
+
+    for (int64_t i = 0; i < len; ++i) {
+      dgl_id_t src_id = all_edges_src_[eid_data[i]];
+      dgl_id_t dst_id = all_edges_dst_[eid_data[i]];
```
Comment (jermainewang) [inline src/graph/graph.cc]: const

File: src/graph/graph.cc [hunk 10]
```diff
@@ -467,37 +467,56 @@ Subgraph Graph::VertexSubgraph(IdArray vids) const {
...
-    dgl_id_t dst_id = all_edges_dst_[eid_data[i]];
-    rst.graph->AddEdge(oldv2newv[src_id], oldv2newv[dst_id]);
-  }
+    rst.graph = std::make_shared<Graph>(IsMultigraph());
+    rst.induced_edges = eids;
+    rst.graph->AddVertices(nodes.size());
+
+    for (int64_t i = 0; i < len; ++i) {
+      dgl_id_t src_id = all_edges_src_[eid_data[i]];
+      dgl_id_t dst_id = all_edges_dst_[eid_data[i]];
```
Comment (jermainewang) [inline src/graph/graph.cc]: const

File: src/graph/graph.cc [hunk 11]
```diff
@@ -467,37 +467,56 @@ Subgraph Graph::VertexSubgraph(IdArray vids) const {
...
 
-  rst.induced_vertices = IdArray::Empty(
-      {static_cast<int64_t>(nodes.size())}, eids->dtype, eids->ctx);
-  std::copy(nodes.begin(), nodes.end(), static_cast<int64_t*>(rst.induced_vertices->data));
+    for (int64_t i = 0; i < NumVertices(); ++i)
+      nodes.push_back(i);
+
+    rst.induced_vertices = IdArray::Empty(
+        {static_cast<int64_t>(nodes.size())}, eids->dtype, eids->ctx);
+    std::copy(nodes.begin(), nodes.end(), static_cast<int64_t*>(rst.induced_vertices->data));
```
Comment (jermainewang) [inline src/graph/graph.cc]: Could you help refactor this function by using the new IdArray routines defined [here](https://github.com/dmlc/dgl/blob/master/include/dgl/array.h)? such as NewIdArray, VecToIdArray, etc. Thx!

File: src/graph/immutable_graph.cc [hunk 12]
```diff
@@ -6,6 +6,7 @@
 
 #include <dgl/immutable_graph.h>
 #include <string.h>
+#include <cassert>
```
Comment (jermainewang) [inline src/graph/immutable_graph.cc]: remove

File: src/graph/immutable_graph.cc [hunk 13]
```diff
@@ -549,7 +550,8 @@ COO::EdgeArray COO::Edges(const std::string &order) const {
   return EdgeArray{src_, dst_, rst_eid};
 }
 
-Subgraph COO::EdgeSubgraph(IdArray eids) const {
+Subgraph COO::EdgeSubgraph(IdArray eids, bool preserve_nodes) const {
+  assert(!preserve_nodes); // TODO(zihao): handle the case of preserve_nodes == true
```
Comment (jermainewang) [inline src/graph/immutable_graph.cc]: Could you help implement this? It shouldn't be too hard. Just create a COO with the given src,dst id and the same number of nodes.

File: src/graph/immutable_graph.cc [hunk 14]
```diff
@@ -696,7 +698,8 @@ Subgraph ImmutableGraph::VertexSubgraph(IdArray vids) const {
                   sg.induced_vertices, sg.induced_edges};
 }
 
-Subgraph ImmutableGraph::EdgeSubgraph(IdArray eids) const {
+Subgraph ImmutableGraph::EdgeSubgraph(IdArray eids, bool preserve_nodes) const {
+  assert(!preserve_nodes); // TODO(zihao): handle the case of preserve_nodes == true
```
Comment (jermainewang) [inline src/graph/immutable_graph.cc]: After implemented COO::EdgeSubgraph, pls remove this.

**Comment by yzh119** [comment]
@jermainewang , I've updated the code according to your review.
As for the refactor of IdArray routines, I found that many parts in the code follow the old style. How about leaving it to the next PR?

**Comment by jermainewang** [comment]
Cool, LGTM!

**Pull request merged**